File: clients/algoliasearch-client-csharp/algoliasearch/Exceptions/AlgoliaException.cs
using System;

namespace Algolia.Search.Exceptions;

/// <summary>
/// Exception thrown when an error occurs in the Algolia client.
/// </summary>
public class AlgoliaException : Exception
{
  /// <summary>
  /// The Correlation-ID header of the response whose handling failed, when
  /// present. Quote it when contacting Algolia support.
  /// </summary>
  public string CorrelationId { get; internal set; }

  /// <summary>
  /// Create a new Algolia exception.
  /// </summary>
  /// <param name="message">The exception details.</param>
  public AlgoliaException(string message)
    : base(message) { }

  /// <summary>
  /// Create a new Algolia exception, with an inner exception.
  /// </summary>
  /// <param name="message"></param>
  /// <param name="inner"></param>
  public AlgoliaException(string message, Exception inner)
    : base(message, inner) { }

  /// <summary>
  /// Appends the Correlation-ID to the formatted string when one is known.
  /// </summary>
  public override string ToString() =>
    CorrelationId == null
      ? base.ToString()
      : $"{base.ToString()} (Correlation-ID: {CorrelationId})";
}

File: clients/algoliasearch-client-php/lib/Exceptions/AlgoliaException.php
<?php

namespace Algolia\AlgoliaSearch\Exceptions;

class AlgoliaException extends \Exception
{
    /**
     * @var null|string
     */
    protected $correlationId;

    /**
     * @param string          $message
     * @param int             $code
     * @param null|\Throwable $previous
     * @param null|string     $correlationId the `Correlation-ID` returned by the failing response
     */
    public function __construct($message = '', $code = 0, $previous = null, $correlationId = null)
    {
        $this->correlationId = $correlationId;

        if (null !== $correlationId && '' !== $correlationId) {
            $message .= ' (Correlation-ID: '.$correlationId.')';
        }

        parent::__construct($message, $code, $previous);
    }

    /**
     * @return null|string the `Correlation-ID` of the failing response, null when the server sent none
     */
    public function getCorrelationId()
    {
        return $this->correlationId;
    }
}
